File: src/camera.cpp
/// @file camera.cpp
/// @brief Implementation of orbit camera.

#include "camera.h"

#include <cmath>

namespace hatching {

Camera::Camera()
    : target_(Eigen::Vector3d::Zero()),
      distance_(3.0),
      theta_(0.0),
      phi_(0.3) {}

void Camera::look_at(const Eigen::Vector3d& target, double distance) {
    target_ = target;
    distance_ = distance;
}

void Camera::rotate(double dx, double dy) {
    theta_ -= dx * 0.01;
    phi_ += dy * 0.01;
    // Clamp elevation to avoid gimbal lock.
    phi_ = std::max(-M_PI / 2.0 + 0.01, std::min(M_PI / 2.0 - 0.01, phi_));
}

void Camera::zoom(double delta) {
    distance_ *= std::exp(-delta * 0.1);
    distance_ = std::max(0.01, distance_);
}

void Camera::pan(double dx, double dy) {
    // Pan in the camera's local XY plane.
    double scale = distance_ * 0.002;
    Eigen::Vector3d right(std::cos(theta_), 0, -std::sin(theta_));
    Eigen::Vector3d up(0, 1, 0);
    target_ -= right * dx * scale;
    target_ += up * dy * scale;
}

Eigen::Vector3d Camera::position() const {
    double x = distance_ * std::cos(phi_) * std::sin(theta_);
    double y = distance_ * std::sin(phi_);
    double z = distance_ * std::cos(phi_) * std::cos(theta_);
    return target_ + Eigen::Vector3d(x, y, z);
}

Eigen::Matrix4f Camera::view_matrix() const {
    Eigen::Vector3d eye = position();
    Eigen::Vector3d forward = (target_ - eye).normalized();
    Eigen::Vector3d world_up(0, 1, 0);
    Eigen::Vector3d right = forward.cross(world_up).normalized();
    Eigen::Vector3d up = right.cross(forward).normalized();

    Eigen::Matrix4f view = Eigen::Matrix4f::Identity();
    view(0, 0) = static_cast<float>(right.x());
    view(0, 1) = static_cast<float>(right.y());
    view(0, 2) = static_cast<float>(right.z());
    view(0, 3) = static_cast<float>(-right.dot(eye));
    view(1, 0) = static_cast<float>(up.x());
    view(1, 1) = static_cast<float>(up.y());
    view(1, 2) = static_cast<float>(up.z());
    view(1, 3) = static_cast<float>(-up.dot(eye));
    view(2, 0) = static_cast<float>(-forward.x());
    view(2, 1) = static_cast<float>(-forward.y());
    view(2, 2) = static_cast<float>(-forward.z());
    view(2, 3) = static_cast<float>(forward.dot(eye));
    return view;
}

Eigen::Matrix4f Camera::projection_matrix(float aspect) const {
    float f = 1.0f / std::tan(fov * static_cast<float>(M_PI) / 360.0f);
    Eigen::Matrix4f proj = Eigen::Matrix4f::Zero();
    proj(0, 0) = f / aspect;
    proj(1, 1) = f;
    proj(2, 2) = (far_plane + near_plane) / (near_plane - far_plane);
    proj(2, 3) = (2.0f * far_plane * near_plane) / (near_plane - far_plane);
    proj(3, 2) = -1.0f;
    return proj;
}

} // namespace hatching

File: web/js/camera.js
/// Orbit camera — matches the C++ Camera class behavior.

export class Camera {
    constructor() {
        this.target = [0, 0, 0];
        this.distance = 3.0;
        this.theta = 0.0;   // azimuth
        this.phi = 0.3;     // elevation
        this.fov = 45.0;    // degrees
        this.near = 0.01;
        this.far = 1000.0;
    }

    lookAt(center, distance) {
        this.target = [center[0], center[1], center[2]];
        this.distance = distance;
    }

    rotate(dx, dy) {
        this.theta -= dx * 0.01;
        this.phi += dy * 0.01;
        const limit = Math.PI / 2 - 0.01;
        this.phi = Math.max(-limit, Math.min(limit, this.phi));
    }

    zoom(delta) {
        this.distance *= Math.exp(-delta * 0.1);
        this.distance = Math.max(0.01, this.distance);
    }

    pan(dx, dy) {
        const scale = this.distance * 0.002;
        const right = [Math.cos(this.theta), 0, -Math.sin(this.theta)];
        const up = [0, 1, 0];
        for (let i = 0; i < 3; i++) {
            this.target[i] -= right[i] * dx * scale;
            this.target[i] += up[i] * dy * scale;
        }
    }

    position() {
        const cp = Math.cos(this.phi), sp = Math.sin(this.phi);
        const ct = Math.cos(this.theta), st = Math.sin(this.theta);
        return [
            this.target[0] + this.distance * cp * st,
            this.target[1] + this.distance * sp,
            this.target[2] + this.distance * cp * ct,
        ];
    }

    viewMatrix() {
        const eye = this.position();
        const fx = this.target[0] - eye[0];
        const fy = this.target[1] - eye[1];
        const fz = this.target[2] - eye[2];
        const fl = Math.sqrt(fx * fx + fy * fy + fz * fz);
        const fwd = [fx / fl, fy / fl, fz / fl];

        // right = forward x world_up
        let rx = fwd[1] * 0 - fwd[2] * 1;
        let ry = fwd[2] * 0 - fwd[0] * 0;
        let rz = fwd[0] * 1 - fwd[1] * 0;
        // world_up = (0,1,0), so: right = (fwd.y*0 - fwd.z*1, fwd.z*0 - fwd.x*0, fwd.x*1 - fwd.y*0)
        // Simplify:
        rx = -fwd[2];
        ry = 0;
        rz = fwd[0];
        const rl = Math.sqrt(rx * rx + ry * ry + rz * rz);
        rx /= rl; ry /= rl; rz /= rl;

        // up = right x forward
        const ux = ry * fwd[2] - rz * fwd[1];
        const uy = rz * fwd[0] - rx * fwd[2];
        const uz = rx * fwd[1] - ry * fwd[0];

        // Column-major for WebGL
        return new Float32Array([
            rx, ux, -fwd[0], 0,
            ry, uy, -fwd[1], 0,
            rz, uz, -fwd[2], 0,
            -(rx * eye[0] + ry * eye[1] + rz * eye[2]),
            -(ux * eye[0] + uy * eye[1] + uz * eye[2]),
            (fwd[0] * eye[0] + fwd[1] * eye[1] + fwd[2] * eye[2]),
            1,
        ]);
    }

    projectionMatrix(aspect) {
        const f = 1.0 / Math.tan((this.fov * Math.PI / 180) / 2);
        const nf = this.near - this.far;
        // Column-major
        return new Float32Array([
            f / aspect, 0, 0, 0,
            0, f, 0, 0,
            0, 0, (this.far + this.near) / nf, -1,
            0, 0, (2 * this.far * this.near) / nf, 0,
        ]);
    }
}
